prow/cmd/exporter/collector.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+type lister interface {
+	List(selector labels.Selector) ([]*prowapi.ProwJob, error)
+}
+
+//https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector
+type prowJobCollector struct {
+	name   string
+	lister lister
+}
+
+func (pjc prowJobCollector) Describe(ch chan<- *prometheus.Desc) {
+	//prometheus.DescribeByCollect(pjc, ch)
+	// Normally, we'd send descriptors into the channel. However, we cannot do so for these
+	// metrics as their label sets are dynamic. This is a take-our-own-risk action and also a
+	// compromise for implementing a metric with both dynamic keys and dynamic values in
+	// the label set.
+	// https://godoc.org/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics
+}
+
+func (pjc prowJobCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
Is it possible to at least have a basic test for this that demonstrate "this prowjob -> this metric"?

Reply:
I will add tests on this function.
In production, it is really the prometheus client that calls this function and check the descriptor (which we disabled).

Reply:
@alvaroaleman 
I did give the unit test a try but ... the challenging part is ...

https://github.com/kubernetes/test-infra/blob/206c830e5c3be9d085ba8d693ce46f755dbc1aec/prow/cmd/exporter/collector.go#L86

The returned `constMetric` is not very test friendly (very private):

```
type constMetric struct {
	desc       *Desc
	valType    ValueType
	val        float64
	labelPairs []*dto.LabelPair
}
```

I cannot easily do the comparison between `actual` and `expected`.

The `Collect` function itself is not that hard to read. I think the non-intuitive part is our non-standard way to use prometheus lib. I hope the links to the prometheus doc are helping.

From the screenshot, we know the `Collect` function works as expected, together with prometheus server. Would it be alright without unit testing code?

Reply:
Well, `MustNewConstMetric` returns a `Metric` and `Meric` in turn is an interface that has this func: 

```
Write(*dto.Metric) error
```

So you could in your test read from the channel, construct a `*dto.Metric`, call `Write` on the metric you got from the channel and then check on the `*dto.Metric.Label`.

Admittedly this is a bit convoluted. Merging this PR as-is and creating an issue for a test is probably fine.

Reply:
Created a follow up: https://github.com/kubernetes/test-infra/pull/14618
Would like to get the current one merged as soon as possible so that I can work on the exporter's deployment and the alerts.